Sessions changes (#87)
Qiskit/documentation: docs/run/sessions.mdx

See [What is the maximum execution time for a Qiskit Runtime job?](max-execution-time) to determine the system limit and
the `max_execution_time` for primitive programs.

> To find the maximum session timeout value for a session, follow the instructions in [Determine session details](#session-details).

Reviewer: ```suggestion
To find the maximum session timeout value for a session, follow the instructions in [Determine session details](run-jobs-in-session#determine-session-details).
```

sessions. Jobs from this new session would then take priority until it
is deactivated or closed, at which point normal job selection resumes.

To find the interactive timeout value for a session, follow the instructions in [Determine session details](#session-details).

<span id="ends"></span>
## What happens when a session ends 

> A session ends by reaching its maximum timeout value,  when it is [closed](run-sessions#close), or when it is cancelled by using the `session.cancel()` method. What happens to unfinished session jobs when the session ends depends on how it ended:

Reviewer: ```suggestion
A session ends by reaching its maximum timeout value,  when it is [closed](run-jobs-in-sessions#close), or when it is canceled by using the `session.cancel()` method. What happens to unfinished session jobs when the session ends depends on how it ended:
```

> A session ends by reaching its maximum timeout value,  when it is [closed](run-jobs-in-sessions#close), or when it is canceled by using the `session.cancel()` method. What happens to unfinished session jobs when the session ends depends on how it ended:

Reviewer: ```suggestion
A session ends by reaching its maximum timeout value,  when it is [closed](run-jobs-in-session#close), or when it is canceled by using the `session.cancel()` method. What happens to unfinished session jobs when the session ends depends on how it ended:
```

<Admonition type="note" title="Important">

>     Previously, `session.close()` **canceled** the session.  Starting with Qiskit Runtime 0.13, `session.close()` **closes** the session. The `session.cancel()` method was added in Qiskit Runtime 0.13.

Author: updated everywhere

sessions. Jobs from this new session would then take priority until it
is deactivated or closed, at which point normal job selection resumes.

To find the interactive timeout value for a session, follow the instructions in [Determine session details](#session-details).

<span id="session-details"></span>
## Determine session details

You can find details about a session by using the `session.details()` method, from the [Quantum Platform Jobs page](https://quantum-computing.ibm.com/jobs), or from the IBM Cloud Jobs page, which you access from your [Instances page](https://cloud.ibm.com/quantum/instances). From the session details you can determine the [maximum](#max-ttl) and [interactive](#ttl) time to live (TTL) values, its status, whether it's currently accepting jobs, and more. 

Example:

``` python
from qiskit_ibm_runtime import QiskitRuntimeService

service = QiskitRuntimeService()

with Session(service=service, backend="ibmq_qasm_simulator") as session:

>   session.details()

Author: For a code example, should we do something like this then: 

with Session(service=service, backend="ibmq_qasm_simulator") as session:
    estimator = Estimator(options=options)
    job = estimator.run(circuit, observable)
   **session.details()**

>   session.details()

Author: Or should just print specific pieces, like you put in your example below?  But in that case should I still put in the rest of the code, where you submit a job; or will that be obvious?

>   session.details()

Reviewer 2: Yeah the code should show submitting a job and also point out specific fields, e.g. 

```
with Session(service=service, backend="ibmq_qasm_simulator") as session:
    estimator = Estimator()
    job = estimator.run(circuit, observable)
    session_details = session.details()
    print(f"The maximum session timeout is {session_details['max_time']} seconds.")
    print(f"The session interactive timeout is {session_details['interactive_timeout']} seconds.")
    print(f"Full session details: {session_details}")
```

_Alternatively_, we can include a sample output and comment on that, e.g. 

```
with Session(service=service, backend="ibmq_qasm_simulator") as session:
    estimator = Estimator()
    job = estimator.run(circuit, observable)
    print(session.details())

This prints
{
'id': 'cki5d18m3kt305s4pndg',
 'backend_name': 'ibm_algiers',
 'interactive_timeout': 300,  # This is the interactive timeout, in seconds
 'max_time': 28800,           # This is the maximum session timeout, in seconds
 'active_timeout': 28800,
 'state': 'closed',
 'accepting_jobs': True,
 'last_job_started': '2023-10-09T19:37:42.004Z',
 'last_job_completed': '2023-10-09T19:38:10.064Z',
 'started_at': '2023-10-09T19:37:42.004Z',
 'closed_at': '2023-10-09T19:38:39.406Z'
}
```

>   ...

Author: I don't see "max_time" in https://api-dev.quantum-computing.ibm.com/runtime/api#/sessions/GetSessionDetailsController_getSessionDetails

>   ...

Reviewer 2: It's [mapped](https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/session.py#L247) in qiskit-ibm-runtime to match the `max_time` parameter for `Session()`

```

<span id="ends"></span>
## What happens when a session ends 

A session ends by reaching its maximum timeout value or by exiting the context manager. See [Close a session](run-sessions#close) for details. What happens when a session closes depends on how the session ended.

If the maximum timeout value was reached:

>     -   Any jobs that are already running are canceled.

Author: That was my understanding from this statement "The session won't accept more jobs but it will continue to run any queued jobs until they are done or the max time expires (and then the same closed rules apply as above)." From this issue: https://github.ibm.com/IBM-Q-Software/requirements-and-priorities/issues/207#issuecomment-63469033 - but I would also appreciate clarification.

>     -   Any jobs that are already running are canceled.

Reviewer 2: Matt confirmed running jobs don't get cancelled.

>     -   Any jobs that are already running are canceled.

Author: So even if you go over the session max time, all queued jobs will finish, @jyu00 @mriedem ?

>     -   Any jobs that are already running are canceled.

Reviewer 2: Queued jobs will be cancelled, but jobs already running won't.

    -   Any queued jobs remaining in the session are put into a failed state.
    -   No further jobs can be submitted to the session.
    -   The session cannot be reopened.

If the maximum timeout value has not been reached:    

- When using Qiskit 12.3 or later releases:
    -   The session is put in `Not accepting new jobs` status.
    -   Any jobs that are already running continue to run.
    -   Any queued jobs remaining in the session continue to run.
    -   No further jobs can be submitted to the session.
    -   The session cannot be reopened.
    -   After all currently queued jobs are completed, the session is immediately closed and will not run any more jobs.

- When using Qiskit 12.2 or previous releases:
    -   Any jobs that are already running continue to run.
    -   Any queued jobs remaining in the session are put into a failed state.
    -   No further jobs can be submitted to the session.

>     -   The session cannot be reopened.

Author: OK - I'm updating this text but leaving the comment open while awaiting feedback about " Any running jobs continue to run   <-- **need to double check this**"  @mriedem , do you know? 

@jyu00 - does "closed vs canceled" only apply to the new changes, or does it also apply to the old content?

>     -   The session cannot be reopened.

Reviewer 2: > Any running jobs continue to run <-- need to double check this" @mriedem , do you know?

Yeah this is true, running jobs would continue to run. 

> does "closed vs canceled" only apply to the new changes, or does it also apply to the old content?

Only the new changes, in the sense that there was only the "cancel" behavior and not "close" prior to 0.13.

>     -   The session cannot be reopened.

Reviewer 3: Let me state that mapping the accepting_jobs field to "closed" is confusing when the closed state on the session in the server and API is something else. I get the meaning on qiskit though (it's closed to accepting more jobs). Since qiskit provides the abstraction layer that's fine. I'm just stating for the record that it gets confusing when talking to devs that knows how the server side works and closed there means the session is kaput. 😄

>     -   The session cannot be reopened.

Author: @jyu00 , how do you cancel a session?  It looks to me like when you exit the context manager, the session is closed and if you use session.close(), the session is closed. If you cancel the first job, is the session canceled?

>     -   The session cannot be reopened.

Reviewer 2: There is a new `session.cancel()` method. This behaves like the old (pre 0.13.0) `close()` and puts all queued jobs in a failed state.

![Diagram showing which job runs next in the queue.](/images/run/jobs-failing.png 'Figure 4: Job selection diagram')

## Summary

- Jobs within an active session take priority over other queued jobs.
- A session becomes active when its first job starts running.
- A session stays active until one of the following happens:
  - Its maximum timeout value is reached. In this case all queued jobs are cancelled, but running jobs will finish. 
  - Its interactive timeout value is reached. In this case the session is deactivated but can be resumed if another session job starts running. 
  - The session is closed or cancelled. This can be done using the corresponding methods or upon exiting a session context.
- Sessions can be used for iterative or batch execution.

## Next steps

<Admonition type="tip" title="Recommendation">

>     - [Run a job in a session.](run-sessions)

Reviewer: ```suggestion
    - [Run a job in a session.](run-jobs-in-session)
```
